cpp/src/arrow/dataset/scanner.h

```diff
@@ -73,6 +74,8 @@ class ARROW_DS_EXPORT ScanOptions {
   // Projector for reconciling the final RecordBatch to the requested schema.
   RecordBatchProjector projector;
 
+  std::unordered_set<std::string> MaterializedFields() const;
+
```

Comment:
Non blocking: Maybe eventually this should be an accessor for a (maybe lazily initialized) `shared_ptr<unordered_set<string>>`

Reply:
Hmm, that's not compatible with being a non-member though

Reply:
I don't understand what this method does. Can you add a docstring?

Reply:
Also, what's the point of having this return an `unordered_set`? Do you plan to implement unions or intersections of those containers?

Reply:
Because it's used as a lookup structure. The ordering and the number of repeats is not important (and should not be exposed). I can still convert it to std::vector.